src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs

```diff
@@ -59,27 +59,48 @@ public void Dispose()
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            if (Interlocked.CompareExchange(ref this._listenerStatus, ListenerRegistering, ListenerNotRegistered) != ListenerNotRegistered)
+            int previousStatus = Interlocked.CompareExchange(ref this._listenerStatus, ListenerRegistering, ListenerNotRegistered);
+
+            if (previousStatus == ListenerRegistering)
+            {
+                throw new InvalidOperationException("The listener is already starting.");
+            }
+            else if (previousStatus == ListenerRegistered)
             {
-                throw new InvalidOperationException("The listener has already been started or is starting.");
+                throw new InvalidOperationException("The listener has already started.");
             }
 
             this.InitializeHost();
 
             try
             {
-                await this._host.RegisterObserverFactoryAsync(this);
+                await RegisterObserverFactoryAsync();
                 Interlocked.CompareExchange(ref this._listenerStatus, ListenerRegistered, ListenerRegistering);
             }
-            catch (DocumentClientException ex) when (ex.StatusCode == HttpStatusCode.NotFound)
+            catch (Exception ex)
             {
-                // Throw a custom error so that it's easier to decipher.
-                string message = $"Either the source collection '{_monitorCollection.CollectionName}' (in database '{_monitorCollection.DatabaseName}')  or the lease collection '{_leaseCollection.CollectionName}' (in database '{_leaseCollection.DatabaseName}') does not exist. Both collections must exist before the listener starts. To automatically create the lease collection, set '{nameof(CosmosDBTriggerAttribute.CreateLeaseCollectionIfNotExists)}' to 'true'.";
-                this._host = null;
-                throw new InvalidOperationException(message, ex);
+                // Reset to NotRegistered
+                this._listenerStatus = ListenerNotRegistered;
+
+                // Throw a custom error if NotFound.
+                if (ex is DocumentClientException docEx && docEx.StatusCode == HttpStatusCode.NotFound)
+                {
+                    // Throw a custom error so that it's easier to decipher.
+                    string message = $"Either the source collection '{_monitorCollection.CollectionName}' (in database '{_monitorCollection.DatabaseName}')  or the lease collection '{_leaseCollection.CollectionName}' (in database '{_leaseCollection.DatabaseName}') does not exist. Both collections must exist before the listener starts. To automatically create the lease collection, set '{nameof(CosmosDBTriggerAttribute.CreateLeaseCollectionIfNotExists)}' to 'true'.";
+                    this._host = null;
```

Comment:
Add `this._listenerStatus = ListenerNotRegistered;` here too?

Reply:
This is nested now (rather than being an exception filter) -- so the call above will execute for this as well.